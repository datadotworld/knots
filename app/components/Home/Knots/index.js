/*
 * knots
 * Copyright 2018 data.world, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the
 * License.
 *
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied. See the License for the specific language governing
 * permissions and limitations under the License.
 *
 * This product includes software developed at
 * data.world, Inc.(http://data.world/).
 */

// @flow

import React, { Component } from 'react';
import { Link, withRouter } from 'react-router-dom';
import { Alert, Button, Container, Table } from 'reactstrap';
import type { KnotType } from '../../../utils/sharedTypes';

import Knot from './Knot';

type Props = {
  knotsStore: {
    knots: Array<KnotType>,
    knotDeleted: boolean,
    knotError: string,
    knotLoaded: boolean
  },
  dockerInstalled: boolean,
  dockerRunning: boolean,
  history: { push: (path: string) => void },
  deleteKnot: (knot: string) => void,
  downloadKnot: (knot: string) => void,
  getKnots: () => void,
  loadValues: (knot: string) => void,
  loadKnot: (knot: string) => void,
  generateUUID: () => void,
  toggleDelete: (knot: KnotType) => void,
  toggleDownloadDisclaimer: (knot: KnotType) => void
};

class Knots extends Component<Props> {
  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.knotsStore.knotDeleted) {
      this.props.getKnots();
    }
  }

  delete = (knot: { name: string }) => {
    this.props.deleteKnot(knot.name);
  };

  download = (knot: { name: string }) => {
    this.props.downloadKnot(knot.name);
  };

  loadValues = (knot: string) => {
    this.props.loadValues(knot);
  };

  loadKnot = (knot: string) => {
    this.props.loadKnot(knot);
  };

  render() {
    const { knots, knotError, knotLoaded } = this.props.knotsStore;
    const { dockerInstalled, dockerRunning, generateUUID } = this.props;

    if (knotLoaded && !knotError) {
      this.props.history.push('/taps');
    }
    return (
      <Container>
        <p className="h2 mb-3 clearfix">
          Saved knots
          <Link to="/taps">
            <Button
              color="secondary"
              outline
              disabled={!dockerInstalled || !dockerRunning}
              size="sm"
              className="float-right"
              onClick={generateUUID}
            >
              New knot
            </Button>
          </Link>
        </p>
        <Alert
          isOpen={!!knotError}
          color="danger"
          className="d-flex justify-content-between"
        >
          <span className="align-self-center">{knotError}</span>
        </Alert>
        <Table>
          <thead className="thead-light">
            <tr>
              <th className="text-center pr-0" style={{ width: '6em' }}>
                Tap
              </th>
              <th className="px-0 fit" />
              <th className="text-center pl-0" style={{ width: '6em' }}>
                Target
              </th>
              <th>Name</th>
              <th>Last Run</th>
              <th className="fit">Actions</th>
            </tr>
          </thead>
          <tbody>
            {knots.map((knot) => (
              <Knot
                key={knot.name}
                knot={knot}
                delete={this.delete}
                download={this.download}
                loadValues={this.loadValues}
                loadKnot={this.loadKnot}
                dockerInstalled={dockerInstalled}
                dockerRunning={dockerRunning}
                toggleDelete={this.props.toggleDelete}
                toggleDownloadDisclaimer={this.props.toggleDownloadDisclaimer}
                generateUUID={this.props.generateUUID}
              />
            ))}
          </tbody>
        </Table>
      </Container>
    );
  }
}

// $FlowFixMe
export default withRouter(Knots);
