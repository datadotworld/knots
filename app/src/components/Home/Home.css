.Home {
  text-align: center;
}

.login {
  margin-top: 10px;
}
