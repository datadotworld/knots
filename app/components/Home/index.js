/*
 * knots
 * Copyright 2018 data.world, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the
 * License.
 *
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied. See the License for the specific language governing
 * permissions and limitations under the License.
 *
 * This product includes software developed at
 * data.world, Inc.(http://data.world/).
 */

// @flow

import React, { Component } from 'react';
import { Alert, Button, Col, Container, Progress, Row } from 'reactstrap';
import { shell } from 'electron';

import Knots from '../../containers/Knots';
import Header from '../Header';
import Create from './Create';

type Props = {
  verifyDocker: () => void,
  getKnots: () => void,
  knotsStore: {
    detectingDocker: boolean,
    dockerVerified: boolean,
    dockerVersion: string,
    dockerRunning: boolean,
    fetchingKnots: boolean,
    knots: Array<{}>
  },
  resetStore: () => void
};

type State = {
  dockerInstalled: boolean,
  dockerRunning: boolean
};

export default class Home extends Component<Props, State> {
  state = {
    dockerInstalled: false,
    dockerRunning: false
  };

  componentWillMount() {
    // reset the store
    this.props.resetStore();
    this.props.verifyDocker();
    this.props.getKnots();
  }

  componentWillReceiveProps(nextProps: Props) {
    this.setState({
      dockerInstalled: !!nextProps.knotsStore.dockerVersion,
      dockerRunning: nextProps.knotsStore.dockerRunning
    });
  }

  onDismiss = () => {
    this.setState({ dockerInstalled: true, dockerRunning: true });

    // Alert will disappear then reappear after one second
    // if Docker hasn't been installed
    setTimeout(() => {
      this.props.verifyDocker();
    }, 1000);
  };

  openLink = (url: string) => {
    shell.openExternal(url);
  };

  render() {
    const {
      detectingDocker,
      dockerVerified,
      fetchingKnots,
      knots
    } = this.props.knotsStore;

    const { dockerInstalled, dockerRunning } = this.state;

    return (
      <div>
        <Header />
        {detectingDocker &&
          fetchingKnots && (
            <Progress value="100" striped animated className="mt-5" />
          )}
        <Container className="mt-5">
          {dockerVerified && (
            <div>
              <Alert
                isOpen={!dockerInstalled || !dockerRunning}
                color="warning"
              >
                <Container>
                  <Row>
                    <Col xs="8">
                      <h4 className="alert-heading">
                        Hey, Docker.{' '}
                        {dockerInstalled ? 'Wake up!' : 'Where are you?'}
                      </h4>
                      <p className="mb-0">
                        Please make sure that Docker is{' '}
                        {dockerInstalled ? 'running' : 'installed'}
                        &nbsp;and try again.
                        <br />
                        <small>
                          <em>
                            KNOTS uses&nbsp;
                            <a
                              href="https://www.docker.com/community-edition#/overview"
                              className="alert-link"
                            >
                              Docker
                            </a>
                            &nbsp;to run pipelines reliably on various operating
                            systems.
                          </em>
                        </small>
                      </p>
                    </Col>
                    <Col xs="4">
                      <p className="text-right">
                        {dockerInstalled ? (
                          <Button
                            outline
                            color="danger"
                            onClick={() => {
                              this.openLink(
                                'https://docs.docker.com/docker-for-mac/install/#install-and-run-docker-for-mac'
                              );
                            }}
                            className="mr-2"
                          >
                            Start Docker
                          </Button>
                        ) : (
                          <Button
                            outline
                            color="danger"
                            onClick={() => {
                              this.openLink(
                                'https://store.docker.com/editions/community/docker-ce-desktop-mac'
                              );
                            }}
                            className="mr-2"
                          >
                            Install Docker
                          </Button>
                        )}
                        <Button color="danger" onClick={this.onDismiss}>
                          Retry
                        </Button>
                      </p>
                    </Col>
                  </Row>
                </Container>
              </Alert>
              {knots.length > 0 && (
                <Knots
                  dockerInstalled={dockerInstalled}
                  dockerRunning={dockerRunning}
                />
              )}
              {knots.length === 0 && <Create {...this.state} />}
            </div>
          )}
        </Container>
      </div>
    );
  }
}
