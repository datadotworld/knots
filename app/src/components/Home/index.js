import React from 'react';
import { Link } from 'react-router-dom';
import { observer, inject } from 'mobx-react';
import { Button } from 'react-bootstrap';

import Header from '../Header';
import './Home.css';

const Home = () => (
  <div className="Home">
    <Header>My Knots</Header>
    <h1>
      Oops! There is <i>Knothing</i> here
    </h1>
    <p>This app, allows you to configure and download an executable Singer</p>
    <Link to="/taps">
      <Button bsStyle="primary" className="login">
        Get Started
      </Button>
    </Link>
  </div>
);

export default inject()(observer(Home));
