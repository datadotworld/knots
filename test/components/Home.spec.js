import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import Home from '../../app/components/Home';
import Knots from '../../app/components/Home/Knots';

const actions = {
  verifyDocker: () => {},
  resetStore: () => {},
  getKnots: () => {}
};

const sampleKnots = [
  {
    tap: {
      name: 'tap-redshift',
      image: 'dataworld/tap-redshift:1.0.0b8'
    },
    target: {
      name: 'target-datadotworld',
      image: 'dataworld/target-datadotworld:1.0.1'
    },
    name: 'knot1'
  },
  {
    tap: {
      name: 'tap-salesforce',
      image: 'dataworld/tap-salesforce:1.4.14'
    },
    target: {
      name: 'target-stitch',
      image: 'dataworld/target-stitch:1.7.4'
    },
    name: 'knot2'
  }
];

describe('Home component', () => {
  it('Detects when Docker is not installed', () => {
    const knotsStore = {
      detectingDocker: false,
      dockerVerified: true,
      dockerVersion: '',
      dockerRunning: false,
      fetchingKnots: false,
      knots: []
    };

    const component = (
      <Router>
        <Home {...actions} knotsStore={knotsStore} />
      </Router>
    );

    const tree = renderer.create(component).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('Detects when Docker is not running', () => {
    const knotsStore = {
      detectingDocker: false,
      dockerVerified: true,
      dockerVersion: '18.03.1-ce-mac65 (24312)',
      dockerRunning: false,
      fetchingKnots: false,
      knots: []
    };

    const component = (
      <Router>
        <Home {...actions} knotsStore={knotsStore} />
      </Router>
    );

    const tree = renderer.create(component).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('Shows jumbotron when there are no saved knots', () => {
    const knotsStore = {
      detectingDocker: false,
      dockerVerified: true,
      dockerVersion: '18.03.1-ce-mac65 (24312)',
      dockerRunning: true,
      fetchingKnots: false,
      knots: []
    };

    const component = (
      <Router>
        <Home {...actions} knotsStore={knotsStore} />
      </Router>
    );

    const tree = renderer.create(component).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('Shows saved knots', () => {
    const props = {
      knotsStore: {
        knots: sampleKnots,
        knotDeleted: false,
        knotError: false,
        knotLoaded: false
      },
      dockerInstalled: true,
      dockerRunning: true,
      deleteKnot: (knot) => {},
      downloadKnot: (knot) => {},
      getKnots: () => {},
      loadValues: (knot) => {},
      history: { push: (path) => {} }
    };

    const component = (
      <Router>
        <Knots {...props} />
      </Router>
    );

    const tree = renderer.create(component).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('Disables knot actions when Docker is not installed', () => {
    const props = {
      knotsStore: {
        knots: sampleKnots,
        knotDeleted: false,
        knotError: false,
        knotLoaded: false
      },
      dockerInstalled: false,
      dockerRunning: false,
      deleteKnot: (knot) => {},
      downloadKnot: (knot) => {},
      getKnots: () => {},
      loadValues: (knot) => {},
      history: { push: (path) => {} }
    };

    const component = (
      <Router>
        <Knots {...props} />
      </Router>
    );

    const tree = renderer.create(component).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('Disables knot actions when Docker is not running', () => {
    const props = {
      knotsStore: {
        knots: sampleKnots,
        knotDeleted: false,
        knotError: false,
        knotLoaded: false
      },
      dockerInstalled: true,
      dockerRunning: false,
      deleteKnot: (knot) => {},
      downloadKnot: (knot) => {},
      getKnots: () => {},
      loadValues: (knot) => {},
      history: { push: (path) => {} }
    };

    const component = (
      <Router>
        <Knots {...props} />
      </Router>
    );

    const tree = renderer.create(component).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
