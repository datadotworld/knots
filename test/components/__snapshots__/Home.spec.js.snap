// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component Detects when Docker is not installed 1`] = `
<div>
  <nav
    className="mb-5 navbar navbar-dark bg-dark"
  >
    <div
      className="container"
    >
      <a
        className="navbar-brand"
        href="#/"
      >
        <img
          alt="KNOTS logo"
          height={40}
          src="test-file-stub"
          width={40}
        />
      </a>
      <ul
        className="navbar-nav nav-pills"
      >
        <li
          className="nav-item"
        >
          <a
            aria-current="page"
            className="nav-link d-none"
            href="/"
            onClick={[Function]}
            style={Object {}}
          >
            Back to saved knots
          </a>
        </li>
      </ul>
    </div>
  </nav>
  <div
    className="mt-5 container"
  >
    <div>
      <div
        className="alert alert-warning fade"
        role="alert"
      >
        <div
          className="d-flex justify-content-between"
        >
          <p
            className="align-self-center mb-0"
          >
            <strong>
              Where are you, Docker?
            </strong>
              Please make sure that Docker is
             
            installed
             and try again.
            <br />
            <small>
              KNOTS uses 
              <a
                className="alert-link"
                href="#"
                onClick={[Function]}
              >
                Docker
              </a>
               to run pipelines reliably on various operating systems.
            </small>
          </p>
          <div
            className="align-self-center"
          >
            <button
              className="mr-2 btn btn-outline-warning"
              onClick={[Function]}
              type="button"
            >
              Install Docker
            </button>
            <button
              className="btn btn-warning"
              onClick={[Function]}
              type="button"
            >
              Retry
            </button>
          </div>
        </div>
      </div>
      <div
        className="jumbotron"
      >
        <h1
          className="display-4"
        >
          There is 
          <em>
            knothing
          </em>
           here yet.
        </h1>
        <p
          className="lead"
        >
          KNOTS allows you to set up simple data pipelines leveraging
           
          <a
            href="#"
            onClick={[Function]}
          >
            Singer
          </a>
           open-source technology.
        </p>
        <a
          href="/taps"
          onClick={[Function]}
        >
          <button
            className="mt-3 btn btn-primary disabled"
            disabled={true}
            onClick={[Function]}
            type={undefined}
          >
            Get Started
          </button>
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Home component Detects when Docker is not running 1`] = `
<div>
  <nav
    className="mb-5 navbar navbar-dark bg-dark"
  >
    <div
      className="container"
    >
      <a
        className="navbar-brand"
        href="#/"
      >
        <img
          alt="KNOTS logo"
          height={40}
          src="test-file-stub"
          width={40}
        />
      </a>
      <ul
        className="navbar-nav nav-pills"
      >
        <li
          className="nav-item"
        >
          <a
            aria-current="page"
            className="nav-link d-none"
            href="/"
            onClick={[Function]}
            style={Object {}}
          >
            Back to saved knots
          </a>
        </li>
      </ul>
    </div>
  </nav>
  <div
    className="mt-5 container"
  >
    <div>
      <div
        className="alert alert-warning fade"
        role="alert"
      >
        <div
          className="d-flex justify-content-between"
        >
          <p
            className="align-self-center mb-0"
          >
            <strong>
              Wake up, Docker!
            </strong>
              Please make sure that Docker is
             
            running
             and try again.
            <br />
            <small>
              KNOTS uses 
              <a
                className="alert-link"
                href="#"
                onClick={[Function]}
              >
                Docker
              </a>
               to run pipelines reliably on various operating systems.
            </small>
          </p>
          <div
            className="align-self-center"
          >
            <button
              className="mr-2 btn btn-outline-warning"
              onClick={[Function]}
              type="button"
            >
              Start Docker
            </button>
            <button
              className="btn btn-warning"
              onClick={[Function]}
              type="button"
            >
              Retry
            </button>
          </div>
        </div>
      </div>
      <div
        className="jumbotron"
      >
        <h1
          className="display-4"
        >
          There is 
          <em>
            knothing
          </em>
           here yet.
        </h1>
        <p
          className="lead"
        >
          KNOTS allows you to set up simple data pipelines leveraging
           
          <a
            href="#"
            onClick={[Function]}
          >
            Singer
          </a>
           open-source technology.
        </p>
        <a
          href="/taps"
          onClick={[Function]}
        >
          <button
            className="mt-3 btn btn-primary disabled"
            disabled={true}
            onClick={[Function]}
            type={undefined}
          >
            Get Started
          </button>
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Home component Disables knot actions when Docker is not installed 1`] = `
<div
  className="container"
>
  <p
    className="h2 mb-3 clearfix"
  >
    Saved knots
    <a
      href="/taps"
      onClick={[Function]}
    >
      <button
        className="float-right btn btn-outline-secondary btn-sm disabled"
        disabled={true}
        onClick={[Function]}
        type={undefined}
      >
        New knot
      </button>
    </a>
  </p>
  <table
    className="table"
  >
    <thead
      className="thead-light"
    >
      <tr>
        <th
          className="text-center pr-0"
          style={
            Object {
              "width": "6em",
            }
          }
        >
          Tap
        </th>
        <th
          className="px-0 fit"
        />
        <th
          className="text-center pl-0"
          style={
            Object {
              "width": "6em",
            }
          }
        >
          Target
        </th>
        <th>
          Name
        </th>
        <th>
          Last Run
        </th>
        <th
          className="fit"
        >
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="align-middle text-center pr-0"
        >
          <img
            alt="tap-redshift logo"
            className="logo"
            src="redshift"
          />
        </td>
        <td
          className="align-middle px-0 text-muted text-center"
        >
          <span
            className="oi oi-chevron-right"
          />
        </td>
        <td
          className="align-middle text-center pr-0"
        >
          <img
            alt="target-datadotworld logo"
            className="logo"
            src="datadotworld"
          />
        </td>
        <th
          className="align-middle"
        >
          knot1
        </th>
        <td
          className="align-middle"
        >
          a few seconds ago
        </td>
        <td
          className="fit align-middle"
        >
          <div
            className="btn-group"
            role="group"
          >
            <button
              className="btn btn-link btn btn-link disabled"
              data-placement="top"
              data-toggle="tooltip"
              disabled={true}
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Sync new data"
              type="button"
            >
              <span
                className="oi oi-media-play"
              />
            </button>
            <button
              className="btn btn-link-secondary btn btn-link-secondary disabled"
              data-placement="top"
              data-toggle="tooltip"
              disabled={true}
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Sync all data"
              type="button"
            >
              <span
                className="oi oi-reload"
              />
            </button>
            <button
              className="btn btn-link-secondary disabled"
              disabled={true}
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Edit"
              type="button"
            >
              <span
                className="oi oi-pencil"
              />
            </button>
            <button
              className="btn btn-link-secondary"
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Export"
              type="button"
            >
              <span
                className="oi oi-cloud-download"
              />
            </button>
            <button
              className="btn btn-link-secondary"
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Delete"
              type="button"
            >
              <span
                className="oi oi-trash"
              />
            </button>
          </div>
        </td>
      </tr>
      <tr>
        <td
          className="align-middle text-center pr-0"
        >
          <img
            alt="tap-salesforce logo"
            className="logo"
            src="salesforce"
          />
        </td>
        <td
          className="align-middle px-0 text-muted text-center"
        >
          <span
            className="oi oi-chevron-right"
          />
        </td>
        <td
          className="align-middle text-center pr-0"
        >
          <img
            alt="target-stitch logo"
            className="logo"
            src="stitch"
          />
        </td>
        <th
          className="align-middle"
        >
          knot2
        </th>
        <td
          className="align-middle"
        >
          a few seconds ago
        </td>
        <td
          className="fit align-middle"
        >
          <div
            className="btn-group"
            role="group"
          >
            <button
              className="btn btn-link btn btn-link disabled"
              data-placement="top"
              data-toggle="tooltip"
              disabled={true}
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Sync new data"
              type="button"
            >
              <span
                className="oi oi-media-play"
              />
            </button>
            <button
              className="btn btn-link-secondary btn btn-link-secondary disabled"
              data-placement="top"
              data-toggle="tooltip"
              disabled={true}
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Sync all data"
              type="button"
            >
              <span
                className="oi oi-reload"
              />
            </button>
            <button
              className="btn btn-link-secondary disabled"
              disabled={true}
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Edit"
              type="button"
            >
              <span
                className="oi oi-pencil"
              />
            </button>
            <button
              className="btn btn-link-secondary"
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Export"
              type="button"
            >
              <span
                className="oi oi-cloud-download"
              />
            </button>
            <button
              className="btn btn-link-secondary"
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Delete"
              type="button"
            >
              <span
                className="oi oi-trash"
              />
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Home component Disables knot actions when Docker is not running 1`] = `
<div
  className="container"
>
  <p
    className="h2 mb-3 clearfix"
  >
    Saved knots
    <a
      href="/taps"
      onClick={[Function]}
    >
      <button
        className="float-right btn btn-outline-secondary btn-sm disabled"
        disabled={true}
        onClick={[Function]}
        type={undefined}
      >
        New knot
      </button>
    </a>
  </p>
  <table
    className="table"
  >
    <thead
      className="thead-light"
    >
      <tr>
        <th
          className="text-center pr-0"
          style={
            Object {
              "width": "6em",
            }
          }
        >
          Tap
        </th>
        <th
          className="px-0 fit"
        />
        <th
          className="text-center pl-0"
          style={
            Object {
              "width": "6em",
            }
          }
        >
          Target
        </th>
        <th>
          Name
        </th>
        <th>
          Last Run
        </th>
        <th
          className="fit"
        >
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="align-middle text-center pr-0"
        >
          <img
            alt="tap-redshift logo"
            className="logo"
            src="redshift"
          />
        </td>
        <td
          className="align-middle px-0 text-muted text-center"
        >
          <span
            className="oi oi-chevron-right"
          />
        </td>
        <td
          className="align-middle text-center pr-0"
        >
          <img
            alt="target-datadotworld logo"
            className="logo"
            src="datadotworld"
          />
        </td>
        <th
          className="align-middle"
        >
          knot1
        </th>
        <td
          className="align-middle"
        >
          a few seconds ago
        </td>
        <td
          className="fit align-middle"
        >
          <div
            className="btn-group"
            role="group"
          >
            <button
              className="btn btn-link btn btn-link disabled"
              data-placement="top"
              data-toggle="tooltip"
              disabled={true}
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Sync new data"
              type="button"
            >
              <span
                className="oi oi-media-play"
              />
            </button>
            <button
              className="btn btn-link-secondary btn btn-link-secondary disabled"
              data-placement="top"
              data-toggle="tooltip"
              disabled={true}
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Sync all data"
              type="button"
            >
              <span
                className="oi oi-reload"
              />
            </button>
            <button
              className="btn btn-link-secondary disabled"
              disabled={true}
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Edit"
              type="button"
            >
              <span
                className="oi oi-pencil"
              />
            </button>
            <button
              className="btn btn-link-secondary"
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Export"
              type="button"
            >
              <span
                className="oi oi-cloud-download"
              />
            </button>
            <button
              className="btn btn-link-secondary"
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Delete"
              type="button"
            >
              <span
                className="oi oi-trash"
              />
            </button>
          </div>
        </td>
      </tr>
      <tr>
        <td
          className="align-middle text-center pr-0"
        >
          <img
            alt="tap-salesforce logo"
            className="logo"
            src="salesforce"
          />
        </td>
        <td
          className="align-middle px-0 text-muted text-center"
        >
          <span
            className="oi oi-chevron-right"
          />
        </td>
        <td
          className="align-middle text-center pr-0"
        >
          <img
            alt="target-stitch logo"
            className="logo"
            src="stitch"
          />
        </td>
        <th
          className="align-middle"
        >
          knot2
        </th>
        <td
          className="align-middle"
        >
          a few seconds ago
        </td>
        <td
          className="fit align-middle"
        >
          <div
            className="btn-group"
            role="group"
          >
            <button
              className="btn btn-link btn btn-link disabled"
              data-placement="top"
              data-toggle="tooltip"
              disabled={true}
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Sync new data"
              type="button"
            >
              <span
                className="oi oi-media-play"
              />
            </button>
            <button
              className="btn btn-link-secondary btn btn-link-secondary disabled"
              data-placement="top"
              data-toggle="tooltip"
              disabled={true}
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Sync all data"
              type="button"
            >
              <span
                className="oi oi-reload"
              />
            </button>
            <button
              className="btn btn-link-secondary disabled"
              disabled={true}
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Edit"
              type="button"
            >
              <span
                className="oi oi-pencil"
              />
            </button>
            <button
              className="btn btn-link-secondary"
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Export"
              type="button"
            >
              <span
                className="oi oi-cloud-download"
              />
            </button>
            <button
              className="btn btn-link-secondary"
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Delete"
              type="button"
            >
              <span
                className="oi oi-trash"
              />
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Home component Shows jumbotron when there are no saved knots 1`] = `
<div>
  <nav
    className="mb-5 navbar navbar-dark bg-dark"
  >
    <div
      className="container"
    >
      <a
        className="navbar-brand"
        href="#/"
      >
        <img
          alt="KNOTS logo"
          height={40}
          src="test-file-stub"
          width={40}
        />
      </a>
      <ul
        className="navbar-nav nav-pills"
      >
        <li
          className="nav-item"
        >
          <a
            aria-current="page"
            className="nav-link d-none"
            href="/"
            onClick={[Function]}
            style={Object {}}
          >
            Back to saved knots
          </a>
        </li>
      </ul>
    </div>
  </nav>
  <div
    className="mt-5 container"
  >
    <div>
      <div
        className="jumbotron"
      >
        <h1
          className="display-4"
        >
          There is 
          <em>
            knothing
          </em>
           here yet.
        </h1>
        <p
          className="lead"
        >
          KNOTS allows you to set up simple data pipelines leveraging
           
          <a
            href="#"
            onClick={[Function]}
          >
            Singer
          </a>
           open-source technology.
        </p>
        <a
          href="/taps"
          onClick={[Function]}
        >
          <button
            className="mt-3 btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type={undefined}
          >
            Get Started
          </button>
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Home component Shows saved knots 1`] = `
<div
  className="container"
>
  <p
    className="h2 mb-3 clearfix"
  >
    Saved knots
    <a
      href="/taps"
      onClick={[Function]}
    >
      <button
        className="float-right btn btn-outline-secondary btn-sm"
        disabled={false}
        onClick={[Function]}
        type={undefined}
      >
        New knot
      </button>
    </a>
  </p>
  <table
    className="table"
  >
    <thead
      className="thead-light"
    >
      <tr>
        <th
          className="text-center pr-0"
          style={
            Object {
              "width": "6em",
            }
          }
        >
          Tap
        </th>
        <th
          className="px-0 fit"
        />
        <th
          className="text-center pl-0"
          style={
            Object {
              "width": "6em",
            }
          }
        >
          Target
        </th>
        <th>
          Name
        </th>
        <th>
          Last Run
        </th>
        <th
          className="fit"
        >
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="align-middle text-center pr-0"
        >
          <img
            alt="tap-redshift logo"
            className="logo"
            src="redshift"
          />
        </td>
        <td
          className="align-middle px-0 text-muted text-center"
        >
          <span
            className="oi oi-chevron-right"
          />
        </td>
        <td
          className="align-middle text-center pr-0"
        >
          <img
            alt="target-datadotworld logo"
            className="logo"
            src="datadotworld"
          />
        </td>
        <th
          className="align-middle"
        >
          knot1
        </th>
        <td
          className="align-middle"
        >
          a few seconds ago
        </td>
        <td
          className="fit align-middle"
        >
          <div
            className="btn-group"
            role="group"
          >
            <button
              className="btn btn-link btn btn-link"
              data-placement="top"
              data-toggle="tooltip"
              disabled={false}
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Sync new data"
              type="button"
            >
              <span
                className="oi oi-media-play"
              />
            </button>
            <button
              className="btn btn-link-secondary btn btn-link-secondary"
              data-placement="top"
              data-toggle="tooltip"
              disabled={false}
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Sync all data"
              type="button"
            >
              <span
                className="oi oi-reload"
              />
            </button>
            <button
              className="btn btn-link-secondary"
              disabled={false}
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Edit"
              type="button"
            >
              <span
                className="oi oi-pencil"
              />
            </button>
            <button
              className="btn btn-link-secondary"
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Export"
              type="button"
            >
              <span
                className="oi oi-cloud-download"
              />
            </button>
            <button
              className="btn btn-link-secondary"
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Delete"
              type="button"
            >
              <span
                className="oi oi-trash"
              />
            </button>
          </div>
        </td>
      </tr>
      <tr>
        <td
          className="align-middle text-center pr-0"
        >
          <img
            alt="tap-salesforce logo"
            className="logo"
            src="salesforce"
          />
        </td>
        <td
          className="align-middle px-0 text-muted text-center"
        >
          <span
            className="oi oi-chevron-right"
          />
        </td>
        <td
          className="align-middle text-center pr-0"
        >
          <img
            alt="target-stitch logo"
            className="logo"
            src="stitch"
          />
        </td>
        <th
          className="align-middle"
        >
          knot2
        </th>
        <td
          className="align-middle"
        >
          a few seconds ago
        </td>
        <td
          className="fit align-middle"
        >
          <div
            className="btn-group"
            role="group"
          >
            <button
              className="btn btn-link btn btn-link"
              data-placement="top"
              data-toggle="tooltip"
              disabled={false}
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Sync new data"
              type="button"
            >
              <span
                className="oi oi-media-play"
              />
            </button>
            <button
              className="btn btn-link-secondary btn btn-link-secondary"
              data-placement="top"
              data-toggle="tooltip"
              disabled={false}
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Sync all data"
              type="button"
            >
              <span
                className="oi oi-reload"
              />
            </button>
            <button
              className="btn btn-link-secondary"
              disabled={false}
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Edit"
              type="button"
            >
              <span
                className="oi oi-pencil"
              />
            </button>
            <button
              className="btn btn-link-secondary"
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Export"
              type="button"
            >
              <span
                className="oi oi-cloud-download"
              />
            </button>
            <button
              className="btn btn-link-secondary"
              onClick={[Function]}
              style={
                Object {
                  "background": "white",
                }
              }
              title="Delete"
              type="button"
            >
              <span
                className="oi oi-trash"
              />
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
